Various Node Manager Bug Fixes

jacderida opened the pull request:
- 94dbf4ef **fix: adding service user on alpine**

  The Alpine distribution has the `adduser` command rather than `useradd`, so we create the service
  user based on which command is available.

- bf1a9e00 **fix: permit removal of manually removed services**

  A community user manually removed one of his service file definitions and he then discovered that
  the node manager would not allow the service to be removed from its registry.

  These changes permit the manual removal of the service definition and also the data and log
  directories. If they have been removed, we just proceed to remove the service from the node
  manager's registry. This solution has only been tested for Systemd-based services. It would need to
  be expanded and tested for other platforms and service managers.

  It would have been possible to submit a PR for the service manager crate to behave in an idempotent
  way and not error if the service file had been removed, but it seems correct for the service manager
  to error in that case. We can handle the error in our application.

- b4245566 **fix: preclude removed services from ops**

  The `start`, `stop`, `remove` and `upgrade` commands are changed such that services that have been
  removed are not considered when operating over a list of services.

  The `start` command in particular was subject to a bug where it tried to `start` removed services,
  and it prevented it from being used to start all services with one command.

- 2af1efca **chore: refresh node registry before commands**

  The node registry is updated to the best of our ability before we run the `start`, `stop`, `remove`
  or `upgrade` commands.

  Here is an example of why this should apply. If we start a service, the node registry now stores the
  PID for the process that is started by the service. It's possible for a user to kill the process
  outside of both the node manager and the OS service infrastructure. However, our service definitions
  have a restart policy, meaning when the user kills the process, the service infrastructure will
  start another one. In this scenario, the node registry now has a PID for a dead process. So if we
  then use the node manager to try and stop the service, it determines there is no process with the
  dead PID, and declares the service has already been stopped.

  The solution is to refresh the registry and get the most up-to-date status before running any of the
  control commands.

## Description

reviewpad:summary

davidrusu commented on sn_node/src/node/flow_ctrl/cmd_ctrl.rs:
```diff
@@ -45,167 +44,162 @@ const ORDER: Ordering = Ordering::SeqCst;
+    pub(crate) async fn push(&mut self, cmd: Cmd) -> Result<()> {
         self.push_internal(cmd, None).await
     }
+    pub(crate) fn q_len(&self) -> usize {
```
cute, but perhaps just spell queue out fully?

joshuef replied:
ah, this may actually not be needed now. was more of a debug sitch. I'll srt it

davidrusu commented on sn_node/src/node/flow_ctrl/cmd_ctrl.rs:
```diff
@@ -45,167 +44,162 @@ const ORDER: Ordering = Ordering::SeqCst;
     monitoring: RateLimits,
-    stopped: Arc<RwLock<bool>>,
+    stopped: bool,
     pub(crate) dispatcher: Arc<Dispatcher>,
     id_counter: Arc<AtomicUsize>,
```
Do these two still need to be behind an `Arc`? Also I don't think you need the AtomicUsize anymore, you could reduce that to `usize` as well

joshuef replied:
Aye, I wasn't sure if i should dive about more there or keep it slim. I can get those cleaned up

joshuef replied:
atomic can be removed easily, dispatcher is harder due to being pulled into the threaded process_cmd. I think we can look at simplifying that down the line 👍

joshuef commented:
#1502 fixes and advances this. closing this branch

jacderida committed:
fix: adding service user on alpine

The Alpine distribution has the `adduser` command rather than `useradd`, so we create the service
user based on which command is available.

jacderida committed:
fix: permit removal of manually removed services

A community user manually removed one of his service file definitions and he then discovered that
the node manager would not allow the service to be removed from its registry.

These changes permit the manual removal of the service definition and also the data and log
directories. If they have been removed, we just proceed to remove the service from the node
manager's registry. This solution has only been tested for Systemd-based services. It would need to
be expanded and tested for other platforms and service managers.

It would have been possible to submit a PR for the service manager crate to behave in an idempotent
way and not error if the service file had been removed, but it seems correct for the service manager
to error in that case. We can handle the error in our application.

jacderida committed:
fix: preclude removed services from ops

The `start`, `stop`, `remove` and `upgrade` commands are changed such that services that have been
removed are not considered when operating over a list of services.

The `start` command in particular was subject to a bug where it tried to `start` removed services,
and it prevented it from being used to start all services with one command.

jacderida committed:
chore: refresh node registry before commands

The node registry is updated to the best of our ability before we run the `start`, `stop`, `remove`
or `upgrade` commands.

Here is an example of why this should apply. If we start a service, the node registry now stores the
PID for the process that is started by the service. It's possible for a user to kill the process
outside of both the node manager and the OS service infrastructure. However, our service definitions
have a restart policy, meaning when the user kills the process, the service infrastructure will
start another one. In this scenario, the node registry now has a PID for a dead process. So if we
then use the node manager to try and stop the service, it determines there is no process with the
dead PID, and declares the service has already been stopped.

The solution is to refresh the registry and get the most up-to-date status before running any of the
control commands.

JasonPaulGithub committed:
Merge branch 'main' into node-man-community-bugs
